lib/utils/compose.coffee

```diff
@@ -103,24 +103,51 @@ toPosixPath = (systemPath) ->
 	systemPath.replace(new RegExp('\\' + path.sep, 'g'), '/')
 
 exports.tarDirectory = tarDirectory = (dir) ->
+	_ = require('lodash')
 	tar = require('tar-stream')
 	klaw = require('klaw')
 	path = require('path')
 	fs = require('mz/fs')
 	streamToPromise = require('stream-to-promise')
+	minimatch = require('minimatch')
 
 	getFiles = ->
 		streamToPromise(klaw(dir))
 		.filter((item) -> not item.stats.isDirectory())
 		.map((item) -> item.path)
 
+	extractIgnored = (file) ->
+		fs.readFile(file, 'utf8').then (content) ->
+			_(content.split(/\r?\n/))
+			.filter((f) -> !_.isEmpty(f))
+			.map((f) -> path.resolve(dir, f))
+			.value()
+
+	toIgnore = []
 	pack = tar.pack()
 	getFiles(dir)
-	.map (file) ->
-		relPath = path.relative(path.resolve(dir), file)
-		Promise.join relPath, fs.stat(file), fs.readFile(file),
-			(filename, stats, data) ->
-				pack.entry({ name: toPosixPath(filename), size: stats.size, mode: stats.mode }, data)
+	.each (file) ->
+		# Find any files that should be ignored
+		if path.relative(path.resolve(dir), file) in ['.gitignore', '.dockerignore' ]
+			return extractIgnored(file)
+			.then (ignoreGlobs) ->
+				# add the ignored files, and also add the ignore file itself
+				ignoreGlobs.push(file)
```

Comment:
Why do we also include the ignore file? That's not standard behaviour for Docker builds, afaict, and you might want it (maybe if you use git inside your build).